File: DSAA Interactive Dashboard/p14.py
import os
import sys

RESET = "\033[0m"
CYAN = "\033[1;36m"
GREEN = "\033[1;32m"
RED = "\033[1;31m"
AMBER = "\033[1;33m"
PURPLE = "\033[1;35m"

INF = 999

V = 0
graph = []
mst = []
elist = []
mst_ok = False

def print_header(title):
    print(f"{PURPLE}\n==================================================")
    print(f"  ◆ {title}")
    print(f"=================================================={RESET}")

def print_user_guide():
    print(f"{CYAN}\n==================================================")
    print("         CRASH COURSE: HOW TO USE THIS SYSTEM")
    print(f"=================================================={RESET}")
    print(" This engine tracks Minimum Spanning Tree (MST) generations via greedy optimizations:\n")
    print(f"{AMBER} [STEP 1]{RESET} Choice 1 maps vertices. Supply edge sequences as: [Src] [Dst] [Weight].")
    print("           Terminate entry streams safely by giving '-1 -1 -1'.")
    print(f"{AMBER} [STEP 2]{RESET} Run a minimum path optimization track:")
    print(f"   • {GREEN}Prim's Algorithm (4):   {RESET} Grows the tree from a single starting root node, adding the cheapest global connection vertex step-by-step.")
    print(f"   • {GREEN}Kruskal's Algorithm (5):{RESET} Sorts all network edges by raw weight and links independent forest structures via Disjoint Set Union (DSU).")
    print(f"{PURPLE}==================================================\n{RESET}")
    input(f"Press {AMBER}[ENTER]{RESET} to initialize structural optimization matrix...")

def find_set(i, p):
    if p[i] == i:
        return i
    p[i] = find_set(p[i], p)
    return p[i]

def create():
    global V, graph, mst, elist, mst_ok
    print_header("GRAPH INITIALIZATION TOOL")
    try:
        V = int(input(f"{AMBER} [?] Enter number of vertices: {RESET}"))
        mst_ok = False
        elist = []
        
        graph = [[INF] * V for _ in range(V)]
        for i in range(V):
            graph[i][i] = 0
        mst = [[0] * V for _ in range(V)]
        
        print(f"{AMBER}\n [!] Enter edges as: [Origin] [Dest] [Weight]{RESET}")
        print(f"{CYAN}     (Type '-1 -1 -1' to finalize setup)\n{RESET}")
        
        while True:
            line = input("  ➔ ").split()
            if not line:
                continue
            o, d, w = int(line[0]), int(line[1]), int(line[2])
            if o == -1:
                break
            if 0 <= o < V and 0 <= d < V and o != d:
                graph[o][d] = graph[d][o] = w
                elist.append({'u': o, 'v': d, 'w': w})
                print(f"{GREEN}    ▲ Edge Saved: {o} <-> {d} (w: {w}){RESET}")
            else:
                print(f"{RED}    ▼ [!] Invalid node IDs or loop self-connection skipped.{RESET}")
        print(f"{GREEN}\n [■] SUCCESS: Graph captured with {len(elist)} active edges.{RESET}")
    except (ValueError, IndexError):
        print(f"{RED} [!] ERROR: Dynamic processing layer failure.{RESET}")
        V = 0

def add_mst_edge(u, v, w):
    global mst
    print(f"{CYAN}  (Node {u}) <==== [w: {w:2d}] ====> (Node {v}){RESET}")
    mst[u][v] = mst[v][u] = w

def prim():
    global V, graph, mst, mst_ok
    if not V:
        print(f"{RED} [!] EMPTY DATA: Please initialize a graph first.{RESET}")
        return
    print_header("PRIM'S MINIMUM SPANNING TREE")
    mst = [[0] * V for _ in range(V)]
    
    sel = [False] * V
    sel[0] = True
    edges, tot = 0, 0
    
    while edges < V - 1:
        min_w = INF
        x, y = -1, -1
        for i in range(V):
            if sel[i]:
                for j in range(V):
                    if not sel[j] and graph[i][j] < min_w:
                        min_w = graph[i][j]
                        x, y = i, j
        if min_w == INF:
            break
        sel[y] = True
        add_mst_edge(x, y, min_w)
        tot += min_w
        edges += 1
        
    if edges < V - 1:
        print(f"{RED}\n [▼] ANALYSIS FAIL: Network graph is disconnected!{RESET}")
    else:
        print(f"{GREEN}\n [■] MST GENERATION SUCCESS | Total Cost Weight: {tot}{RESET}")
        mst_ok = True

def kruskal():
    global V, mst, elist, mst_ok
    if not V:
        print(f"{RED} [!] EMPTY DATA: Please initialize a graph first.{RESET}")
        return
    print_header("KRUSKAL'S MINIMUM SPANNING TREE")
    mst = [[0] * V for _ in range(V)]
    
    p = list(range(V))
    sorted_edges = sorted(elist, key=lambda edge: edge['w'])
    
    edges, tot = 0, 0
    for edge in sorted_edges:
        if edges >= V - 1:
            break
        ru = find_set(edge['u'], p)
        rv = find_set(edge['v'], p)
        if ru != rv:
            p[ru] = rv
            add_mst_edge(edge['u'], edge['v'], edge['w'])
            tot += edge['w']
            edges += 1
            
    if edges < V - 1:
        print(f"{RED}\n [▼] ANALYSIS FAIL: Network graph is disconnected!{RESET}")
    else:
        print(f"{GREEN}\n [■] MST GENERATION SUCCESS | Total Cost Weight: {tot}{RESET}")
        mst_ok = True

def draw_tree(n, parent, depth, last):
    global V, mst
    for _ in range(depth - 1):
        print("  | ", end="")
    if depth > 0:
        print("  +-- ", end="")
    if parent == -1:
        print(f"{GREEN}[Root Node {n}]{RESET}")
    else:
        print(f"{CYAN}Node {n}{AMBER} (w:{mst[parent][n]}){RESET}")
        
    ch = [i for i in range(V) if mst[n][i] and i != parent]
    for i, child in enumerate(ch):
        draw_tree(child, n, depth + 1, i == len(ch) - 1)

def main():
    os.system('cls' if os.name == 'nt' else 'clear')
    print_user_guide()
    
    print(f"{CYAN}==================================================")
    print("     ■ CRITICAL MST CORE ENGINE & ANALYSIS ■")
    print(f"=================================================={RESET}")
    
    while True:
        print(f"{CYAN}\n [SYSTEM STATUS] Active Vertices: {V} | Active Edges: {len(elist)}{RESET}")
        print("  1.  [+] Create Network Graph")
        print("  2.  [#] Render Adjacency Matrix")
        print("  3.  [*] Render Structural Map")
        print("  4.  [^] Execute Prim's Path Mapping")
        print("  5.  [^] Execute Kruskal's Path Mapping")
        print("  6.  [$] Draw Hierarchical Tree Visual")
        print("  7.  [X] Terminate Execution Engine")
        
        try:
            choice = input(f"{AMBER} >> Select Target Action Option: {RESET}").strip()
            if not choice:
                continue
            ch = int(choice)
        except ValueError:
            print(f"{RED} [!] WARNING: Selection outside variable parameters.{RESET}")
            continue
            
        if ch == 7:
            print(f"{RED}\n [!] SHUTTING DOWN CORE SYSTEM OPERATIONS...\n{RESET}")
            sys.exit(0)
            
        if ch == 1:
            create()
        elif ch in (2, 3):
            if not V:
                print(f"{RED} [!] ACCESS REJECTED: Network matrix layer is currently empty.{RESET}")
                continue
            print_header("MATRIX DATA REPRESENTATION" if ch == 2 else "GRAPH CONSOLIDATION MAP")
            for i in range(V):
                if ch == 3:
                    print(f"{GREEN} (Node {i}) {RESET}", end="")
                for j in range(V):
                    if ch == 2:
                        if graph[i][j] == INF:
                            print(f"{RED} INF {RESET}", end="")
                        else:
                            print(f"{CYAN} {graph[i][j]:3d} {RESET}", end="")
                    elif graph[i][j] and graph[i][j] != INF:
                        print(f"{CYAN} ==({graph[i][j]})==> {RESET}[{j}]", end="")
                print()
        elif ch == 4:
            prim()
        elif ch == 5:
            kruskal()
        elif ch == 6:
            if mst_ok:
                print_header("HIERARCHICAL TREE ARCHITECTURE")
                draw_tree(0, -1, 0, True)
            else:
                print(f"{RED} [!] CRITICAL ERROR: Active MST profile data missing. Run option 4 or 5 first.{RESET}")
        else:
            print(f"{RED} [!] WARNING: Selection outside variable parameters.{RESET}")

if __name__ == "__main__":
    main()

File: DSAA Interactive Dashboard/p14.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// ANSI Color Codes for high-tech look
#define RESET   "\033[0m"
#define CYAN    "\033[1;36m"
#define GREEN   "\033[1;32m"
#define RED     "\033[1;31m"
#define AMBER   "\033[1;33m"
#define PURPLE  "\033[1;35m"

#define MAX 6
#define INF 999

typedef struct { int u, v, w; } Edge;
int graph[MAX][MAX], mst[MAX][MAX], V = 0, E_cnt = 0;
Edge elist[MAX * MAX];
bool mst_ok = false;

void printHeader(char* title) {
    printf(PURPLE "\n==================================================\n");
    printf("  ◆ %s\n", title);
    printf("==================================================\n" RESET);
}

void printUserGuide() {
    printf(CYAN "\n==================================================\n");
    printf("         CRASH COURSE: HOW TO USE THIS SYSTEM\n");
    printf("==================================================\n" RESET);
    printf(" This engine tracks Minimum Spanning Tree (MST) generations via greedy optimizations:\n\n");
    printf(AMBER " [STEP 1]" RESET " Choice 1 maps vertices. Supply edge sequences as: [Src] [Dst] [Weight].\n");
    printf("           Terminate entry streams safely by giving '-1 -1 -1'.\n");
    printf(AMBER " [STEP 2]" RESET " Run a minimum path optimization track:\n");
    printf("   • " GREEN "Prim's Algorithm (4):   " RESET " Grows the tree from a single starting root node, adding the cheapest global connection vertex step-by-step.\n");
    printf("   • " GREEN "Kruskal's Algorithm (5):" RESET " Sorts all network edges by raw weight and links independent forest structures via Disjoint Set Union (DSU).\n");
    printf(PURPLE "==================================================\n\n" RESET);
    printf("Press " AMBER "[ENTER]" RESET " to initialize structural optimization matrix...");
    while(getchar() != '\n'); // Clear buffer
    getchar();
}

int cmp(const void* a, const void* b) { return ((Edge*)a)->w - ((Edge*)b)->w; }
int find(int i, int p[]) { return (p[i] == i) ? i : (p[i] = find(p[i], p)); }

void create() {
    int o, d, w;
    printHeader("GRAPH INITIALIZATION TOOL");
    printf(AMBER " [?] Enter number of vertices (Max %d): " RESET, MAX);
    scanf("%d", &V);
    E_cnt = mst_ok = 0;

    for (int i = 0; i < V; i++) 
        for (int j = 0; j < V; j++) graph[i][j] = (i == j) ? 0 : INF;

    printf(AMBER "\n [!] Enter edges as: [Origin] [Dest] [Weight]\n" RESET);
    printf(CYAN "     (Type '-1 -1 -1' to finalize setup)\n\n" RESET);
    while (1) {
        printf("  ➔ ");
        scanf("%d %d %d", &o, &d, &w);
        if (o == -1) break;
        if (o >= 0 && o < V && d >= 0 && d < V && o != d) {
            graph[o][d] = graph[d][o] = w;
            elist[E_cnt++] = (Edge){o, d, w};
            printf(GREEN "    ▲ Edge Saved: %d <-> %d (w: %d)\n" RESET, o, d, w);
        } else {
            printf(RED "    ▼ [!] Invalid node IDs or loop self-connection skipped.\n" RESET);
        }
    }
    printf(GREEN "\n [■] SUCCESS: Graph captured with %d active edges.\n" RESET, E_cnt);
}

void add_mst_edge(int u, int v, int w, int *edges, int *tot) {
    printf(CYAN "  (Node %d) <==== [w: %2d] ====> (Node %d)\n" RESET, u, w, v);
    mst[u][v] = mst[v][u] = w;
    (*tot) += w; (*edges)++;
}

void prim() {
    if (!V) { printf(RED " [!] EMPTY DATA: Please initialize a graph first.\n" RESET); return; }
    printHeader("PRIM'S MINIMUM SPANNING TREE");
    for (int i = 0; i < V; i++) for (int j = 0; j < V; j++) mst[i][j] = 0;

    bool sel[MAX] = {true};
    int edges = 0, tot = 0;

    while (edges < V - 1) {
        int min = INF, x = -1, y = -1;
        for (int i = 0; i < V; i++) if (sel[i])
            for (int j = 0; j < V; j++)
                if (!sel[j] && graph[i][j] < min) { min = graph[i][j]; x = i; y = j; }
        if (min == INF) break;
        sel[y] = true;
        add_mst_edge(x, y, min, &edges, &tot);
    }
    if (edges < V - 1) printf(RED "\n [▼] ANALYSIS FAIL: Network graph is disconnected!\n" RESET);
    else { printf(GREEN "\n [■] MST GENERATION SUCCESS | Total Cost Weight: %d\n" RESET, tot); mst_ok = true; }
}

void kruskal() {
    if (!V) { printf(RED " [!] EMPTY DATA: Please initialize a graph first.\n" RESET); return; }
    printHeader("KRUSKAL'S MINIMUM SPANNING TREE");
    for (int i = 0; i < V; i++) for (int j = 0; j < V; j++) mst[i][j] = 0;

    int p[MAX], edges = 0, tot = 0;
    for (int i = 0; i < V; i++) p[i] = i;
    qsort(elist, E_cnt, sizeof(Edge), cmp);

    for (int i = 0; i < E_cnt && edges < V - 1; i++) {
        int ru = find(elist[i].u, p), rv = find(elist[i].v, p);
        if (ru != rv) { p[ru] = rv; add_mst_edge(elist[i].u, elist[i].v, elist[i].w, &edges, &tot); }
    }
    if (edges < V - 1) printf(RED "\n [▼] ANALYSIS FAIL: Network graph is disconnected!\n" RESET);
    else { printf(GREEN "\n [■] MST GENERATION SUCCESS | Total Cost Weight: %d\n" RESET, tot); mst_ok = true; }
}

void draw_tree(int n, int parent, int depth, bool last) {
    for (int i = 0; i < depth - 1; i++) printf("  | ");
    if (depth > 0) printf("  +-- ");
    if (parent == -1) printf(GREEN "[Root Node %d]\n" RESET, n); 
    else printf(CYAN "Node %d" AMBER " (w:%d)\n" RESET, n, mst[parent][n]);

    int ch[MAX], c_cnt = 0;
    for (int i = 0; i < V; i++) if (mst[n][i] && i != parent) ch[c_cnt++] = i;
    for (int i = 0; i < c_cnt; i++) draw_tree(ch[i], n, depth + 1, i == c_cnt - 1);
}

int main() {
    int ch;
    #ifdef _WIN32
        system("cls");
    #else
        system("clear");
    #endif

    printUserGuide();

    printf(CYAN "==================================================\n");
    printf("     ■ CRITICAL MST CORE ENGINE & ANALYSIS ■\n");
    printf("==================================================\n" RESET);

    while (1) {
        printf(CYAN "\n [SYSTEM STATUS] Active Vertices: %d/%d | Active Edges: %d\n" RESET, V, MAX, E_cnt);
        printf("  1.  [+] Create Network Graph\n");
        printf("  2.  [#] Render Adjacency Matrix\n");
        printf("  3.  [*] Render Structural Map\n");
        printf("  4.  [^] Execute Prim's Path Mapping\n");
        printf("  5.  [^] Execute Kruskal's Path Mapping\n");
        printf("  6.  [$] Draw Hierarchical Tree Visual\n");
        printf("  7.  [X] Terminate Execution Engine\n");
        printf(AMBER " >> Select Target Action Option: " RESET);
        
        if (scanf("%d", &ch) != 1 || ch == 7) {
            printf(RED "\n [!] SHUTTING DOWN CORE SYSTEM OPERATIONS...\n" RESET);
            break;
        }

        if (ch == 1) create();
        else if (ch == 2 || ch == 3) {
            if (!V) { printf(RED " [!] ACCESS REJECTED: Network matrix layer is currently empty.\n" RESET); continue; }
            printHeader(ch == 2 ? "MATRIX DATA REPRESENTATION" : "GRAPH CONSOLIDATION MAP");
            for (int i = 0; i < V; i++) {
                if (ch == 3) printf(GREEN " (Node %d) " RESET, i);
                for (int j = 0; j < V; j++) {
                    if (ch == 2) {
                        if (graph[i][j] == INF) printf(RED " INF " RESET);
                        else printf(CYAN " %3d " RESET, graph[i][j]);
                    }
                    else if (graph[i][j] && graph[i][j] != INF) printf(CYAN " ==(%d)==> " RESET "[%d]", graph[i][j], j);
                }
                printf("\n");
            }
        } 
        else if (ch == 4) prim();
        else if (ch == 5) kruskal();
        else if (ch == 6) {
            if (mst_ok) { 
                printHeader("HIERARCHICAL TREE ARCHITECTURE"); 
                draw_tree(0, -1, 0, true); 
            } else {
                printf(RED " [!] CRITICAL ERROR: Active MST profile data missing. Run option 4 or 5 first.\n" RESET);
            }
        } else {
            printf(RED " [!] WARNING: Selection outside variable parameters.\n" RESET);
        }
    }
    return 0;
}